File: problem_4/solution.py
def segregate(arr):
    """
    Put negative and 0 value at left size of the array.
    """
    j = 0
    for i, val in enumerate(arr):
        if val <= 0:
            arr[i], arr[j] = arr[j], arr[i] # shift to the left of array
            j += 1
    return j


def find_missing_pos(arr):
    size = len(arr)
    neg_size = segregate(arr)
    pos_size = size - neg_size
    i = neg_size
    while i < size:
        index = abs(arr[i]) - 1 + neg_size
        if (index < size and arr[index] > 0):
            arr[index] = -arr[index]
        i += 1

    i = neg_size
    while i < size:
        if arr[i] > 0:
            return i - neg_size + 1
        i += 1

    return pos_size + 1


if __name__ == '__main__':
    filename = "test_cases.txt"
    with open(filename) as file:
        content = file.readline()
        while content:
            arr, ans = content.split()
            ans = int(ans)
            arr = [int(a) for a in arr.split(',')]
            if(find_missing_pos(arr) == ans):
                print("Passed")
            else:
                print("Failed")
            content = file.readline()

File: problem_4/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func segregate(arr []int) int {
	n := len(arr)
	j := 0
	for i := 0; i < n; i++ {
		if arr[i] <= 0 {
			arr[i], arr[j] = arr[j], arr[i]
			j++
		}
	}
	return j

}

func findMissingPos(arr []int) int {
	size := len(arr)
	neg_size := segregate(arr)
	pos_size := size - neg_size
	for i := neg_size; i < size; i++ {
		index := int(math.Abs(float64(arr[i]))) - 1 + neg_size
		if index < size && arr[index] > 0 {
			arr[index] = -arr[index]
		}
	}

	for i := neg_size; i < size; i++ {
		if arr[i] > 0 {
			return i - neg_size + 1
		}
	}

	return pos_size + 1
}

func main() {
	file, err := os.Open("test_cases.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		// Reading text_cases.txt in variables
		content := strings.Split(scanner.Text(), " ")
		str_arr := strings.Split(content[0], ",")
		arr := []int{}
		for _, i := range str_arr {
			j, err := strconv.Atoi(i)
			if err != nil {
				fmt.Println(err)
			}
			arr = append(arr, j)
		}
		ans, ans_err := strconv.Atoi(content[1])
		if ans_err != nil {
			fmt.Println(ans_err)
		}

		if findMissingPos(arr) == ans {
			fmt.Println("Passed")
		} else {
			fmt.Println("Failed")
		}

		if err := scanner.Err(); err != nil {
			fmt.Println(err)
		}
	}
}
